### ISSUE TITLE
15% perf delta between array access and span access

### ISSUE THREAD

**Issue by stephentoub** [dotnet/coreclr#13097]
Repro:
```C#
using System;
using System.Diagnostics;
using System.Linq;
using System.Numerics;
using System.Runtime.CompilerServices;

class Program
{
    static void Main()
    {
        const int Iters = 1000000;
        var sw = new Stopwatch();
        byte[] array = new byte[1000];
        Span<byte> span = array;
        UseArray(array);
        UseSpan(span);

        while (true)
        {
            sw.Restart();
            for (int i = 0; i < Iters; i++) UseArray(array);
            sw.Stop();
            double arrayTime = sw.Elapsed.TotalMilliseconds;

            sw.Restart();
            for (int i = 0; i < Iters; i++) UseSpan(span);
            sw.Stop();
            double spanTime = sw.Elapsed.TotalMilliseconds;

            Console.WriteLine($"Array: {arrayTime} Span: {spanTime} Diff: {(spanTime - arrayTime)/arrayTime*100:F2}%");
        }
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    static int UseArray(byte[] value)
    {
        int i = value.Length;
        while (i > 0 && value[i - 1] == 0) i--;
        return i;
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    static int UseSpan(Span<byte> value)
    {
        int i = value.Length;
        while (i > 0 && value[i - 1] == 0) i--;
        return i;
    }
}
```
On my machine, the span version is ~15% slower than the array version.  In contrast, if I change the body of the tests to something like:
```C#
int sum = 0;
for (int i = 0; i < value.Length; i++) sum += value[i];
return sum;
```
the difference ends up being much smaller, typically < 1%, and sometimes the span version is even faster.

**Comment by mikedn** [comment]
On my machine the difference is even higher - ~23%. But it's not readily obvious from the generated code what the reason is:
```asm
; array version
G_M55886_IG03:
       418BD0               mov      edx, r8d
       85D2                 test     edx, edx
       7E15                 jle      SHORT G_M55886_IG05
G_M53282_IG04:
       448D42FF             lea      r8d, [rdx-1]
       443BC0               cmp      r8d, eax
       7313                 jae      SHORT G_M55886_IG07
       4D63C8               movsxd   r9, r8d
       42807C091000         cmp      byte  ptr [rcx+r9+16], 0
       74E5                 je       SHORT G_M55886_IG03
; span version
G_M55887_IG03:
       418BC8               mov      ecx, r8d
G_M53282_IG04:
       85C9                 test     ecx, ecx
       7E17                 jle      SHORT G_M55887_IG05
       448D41FF             lea      r8d, [rcx-1]
       458BC8               mov      r9d, r8d
       443BCA               cmp      r9d, edx
       7312                 jae      SHORT G_M55887_IG07
       4D63C9               movsxd   r9, r9d
       42803C0800           cmp      byte  ptr [rax+r9], 0
       74E3                 je       SHORT G_M55887_IG03
```
The range check is not eliminated in either version (because the loop counts down). Loop alignment has no effect. The span version has an extra `mov` in the middle but it should not have a significant impact on performance.

**Comment by mikedn** [comment]
Note that you get slightly better codegen if you change your code like this:
```
int i = value.Length - 1;
while (i >= 0 && value[i] == 0) i--;
return i + 1;
```
In this case both the span and array have the same (good) speed.

I'm inclined to say that this has nothing to do with span, it's just a general case of inability to generate better code.

**Comment by stephentoub** [comment]
Thanks, @mikedn.

> I'm inclined to say that this has nothing to do with span, it's just a general case of inability to generate better code.

Hmm, I'm not quite understanding.  If it's just a general case of an inability to generate better code, wouldn't that also apply to the array case (in the original example)?  My concern here is that we're trying to get to a place where span can be substituted for array without fear of regression, but if cases like this can result in 20% differences, that makes it much more difficult for spans to be substituted in arbitrarily.

> if you change your code like this

Thanks.  This came directly from the BigInteger code, which for this purpose I tried to modify as little as possible, but I will go back and try to clean this up to see if it can alleviate the differences in this and other cases in the ctor it came from.

> But it's not readily obvious from the generated code what the reason is

Yeah, I didn't see anything obvious either.  The two main differences seemed to be the extra mov and the +16 in the cmp.

**Comment by mikedn** [comment]
> Hmm, I'm not quite understanding. If it's just a general case of an inability to generate better code, wouldn't that also apply to the array case (in the original example)?

Well, the obvious problems exists in both the array and span code - the range check is there, the signed index cast (movsxd) is there, one or more unnecessary register moves are there. The code generated would ideally look like below but the JIT can't produce that in either version:
```asm
G_M1720_IG03:
       FFC9                 dec      rcx
G_M1720_IG04:
       85C9                 test     rcx, rcx
       7C0F                 jl       SHORT G_M1720_IG05
       42803C0000           cmp      byte  ptr [rax+rcx], 0  ; +16 if array
       74EC                 je       SHORT G_M1720_IG03
```
Anyway, I have to check what's up with the extra `mov`. But I wouldn't be surprised if the span version is simply unlucky and that some kind of array code will also exhibit this issue. We'll see.

And it's probably worth pointing out where do those extra `mov` come from - from `i - 1` and `i--`. The JIT does CSE so `i - 1` appears only once but in doing so it introduces a temporary variables and some copies.

**Comment by mikedn** [comment]
OK, I moved the "benchmark" to C++/ASM so I can play with the code a bit. It's the `mov` that causes the slowdown. Sort of. 

Any instruction, even a `nop`, has a similar effect. You can add a `nop` somewhere in the array version of a loop and you get the same slowdown. So it's not the instruction itself but something else. Maybe code size but the following experiment indicates that's probably not the case.

On my machine the original code generates:
```
Array: 632.0378 Span: 788.6776 Diff: 24.78%
Array: 629.1638 Span: 786.5573 Diff: 25.02%
Array: 632.8486 Span: 791.0285 Diff: 24.99%
```
Change both loops to `while (i > 0 && value[i - 1] - 42 == 0) i--;` and fill the array with 42. Now we have:
```
Array: 566.4973 Span: 638.3346 Diff: 12.68%
Array: 563.6572 Span: 636.9845 Diff: 13.01%
Array: 565.1912 Span: 637.1021 Diff: 12.72%
```
So a new instruction is added and the code gets larger. Now it's faster and the difference between the array and the loop version is smaller. Very intuitive, good luck measuring code performance :).

**Comment by mikedn** [comment]
Now where does that `mov` come from and what does `Span` to do with it?

The relevant IR looks like:
```
; array
          [000026] ------------  *  STMT      void  (IL 0x00E...0x013)
( 31, 31) [000025] ---XG-------  \--*  JTRUE     void  
(  1,  1) [000023] ------------     |  /--*  CNS_INT   int    0
( 29, 29) [000024] J--XG--N----     \--*  EQ        int   
( 13, 12) [000022] a--XG-------        |  /--*  IND       ubyte 
(  1,  1) [000042] ------------        |  |  |  /--*  CNS_INT   long   16 Fseq[#FirstElem]
( 10,  9) [000043] -------N----        |  |  \--*  ADD       byref 
(  6,  6) [000040] ------------        |  |     |  /--*  CAST      long <- int
(  1,  1) [000037] ------------        |  |     |  |  |  /--*  CNS_INT   int    -1 Fseq[#ConstantIndex]
(  5,  4) [000035] ------------        |  |     |  |  \--*  ADD       int   
(  3,  2) [000036] i-----------        |  |     |  |     \--*  LCL_VAR   int    V01 loc0         
(  9,  8) [000041] -------N----        |  |     \--*  ADD       byref 
(  3,  2) [000034] ------------        |  |        \--*  LCL_VAR   ref    V00 arg0         
( 27, 27) [000044] ---XG-------        \--*  COMMA     ubyte 
( 14, 15) [000039] ---X--------           \--*  ARR_BOUNDS_CHECK_Rng void  
(  1,  1) [000020] ------------              |  /--*  CNS_INT   int    -1
(  5,  4) [000021] ------------              +--*  ADD       int   
(  3,  2) [000019] ------------              |  \--*  LCL_VAR   int    V01 loc0         
(  5,  4) [000038] ---X--------              \--*  ARR_LENGTH int   
(  3,  2) [000018] ------------                 \--*  LCL_VAR   ref    V00 arg0         
; span
          [000027] ------------  *  STMT      void  (IL 0x012...0x01D)
          [000023] -----+------  |     /--*  CNS_INT   int    -1
          [000024] -----+------  |  /--*  ADD       int   
          [000022] -----+------  |  |  \--*  LCL_VAR   int    V01 loc0         
          [000026] -A---+------  \--*  ASG       int   
          [000025] D----+-N----     \--*  LCL_VAR   int    V02 tmp0         

***** BB05, stmt 6
          [000044] ------------  *  STMT      void  (IL   ???...  ???)
          [000043] ---XG+------  \--*  JTRUE     void  
          [000041] -----+------     |  /--*  CNS_INT   int    0
          [000042] J--XG+-N----     \--*  EQ        int   
          [000076] *--XG+------        |  /--*  IND       ubyte 
          [000034] -----+------        |  |  |  /--*  CAST      long <- int
          [000028] -----+------        |  |  |  |  \--*  LCL_VAR   int    V02 tmp0         
          [000038] ----G+------        |  |  \--*  ADD       byref 
          [000037] -----+------        |  |     \--*  LCL_VAR   byref  V03 tmp1         
          [000039] ---XG+------        \--*  COMMA     ubyte 
          [000033] ---XG+------           \--*  ARR_BOUNDS_CHECK_Rng void  
          [000029] -----+------              +--*  LCL_VAR   int    V02 tmp0         
          [000032] -----+------              \--*  LCL_VAR   int    V04 tmp2         
```
In the array case we start with `GT_INDEX` . It gets expanded by morph and in the process `i - 1` is cloned because it has a low cost.
In the span case `get_Item` is expanded during import and `i - 1` is stored in a temporary rather than being cloned.

After this CSE comes in and finds `i - 1` and `i--`. It creates a temporary to hold the result of `i - 1` and replaces all `i - 1` uses with that temporary. In the array case this works fine but in the span case we already had a temporary and now we have two:
```
t1 = i - 1
t2 = t1
a[t2] == 0
```

Copy propagation can't seem to handle `t2 = t1` and that copy becomes our `mov`.

I've seen this issue before, it's really a problem with CSE and copy prop. In this case it happens to be triggered by span specifics. However, this issue can be triggered by other things, including array accesses.

**Comment by stephentoub** [comment]
Thanks for the detailed look, @mikedn!

Here's another example where there appears to be an extra `mov` in the span version (this is again based on a snippet of code from BigInteger):
```C#
using System;
using System.Diagnostics;
using System.Linq;

class Program
{
    static void Main()
    {
        var sw = new Stopwatch();
        var r = new Random(42);

        int[] srcArray = Enumerable.Range(0, 100).Select(_ => r.Next()).ToArray();
        byte[] dstArray = new byte[srcArray.Length * 4];

        Span<int> srcSpan = srcArray;
        Span<byte> dstSpan = dstArray;

        const int Iters = 3000000;

        while (true)
        {
            sw.Restart();
            for (int i = 0; i < Iters; i++) UseArray(srcArray, dstArray);
            sw.Stop();
            Console.WriteLine(sw.Elapsed.TotalMilliseconds);

            sw.Restart();
            for (int i = 0; i < Iters; i++) UseSpan(srcSpan, dstSpan);
            sw.Stop();
            Console.WriteLine(sw.Elapsed.TotalMilliseconds);

            Console.WriteLine();
        }
    }

    static void UseArray(int[] src, byte[] dst)
    {
        int b = 0;
        for (int i = 0; i < src.Length; i++)
        {
            int val = src[i];
            for (int j = 0; j < 4; j++)
            {
                dst[b++] = (byte)val;
                val >>= 8;
            }
        }
    }

    static void UseSpan(Span<int> src, Span<byte> dst)
    {
        int b = 0;
        for (int i = 0; i < src.Length; i++)
        {
            int val = src[i];
            for (int j = 0; j < 4; j++)
            {
                dst[b++] = (byte)val;
                val >>= 8;
            }
        }
    }
}
```
The inner loop for the array version has:
```
G_M49838_IG04:
       8D7801               lea      edi, [rax+1]
       3BC6                 cmp      eax, esi
       732A                 jae      SHORT G_M49838_IG07
       4863C0               movsxd   rax, eax
       410FB6DA             movzx    rbx, r10b
       885C0210             mov      byte  ptr [rdx+rax+16], bl
       41C1FA08             sar      r10d, 8
       41FFC3               inc      r11d
       4183FB04             cmp      r11d, 4
       8BC7                 mov      eax, edi
       7CDF                 jl       SHORT G_M49838_IG04
```
while the inner loop for the span version has:
```
G_M5180_IG04:
       418D7001             lea      esi, [r8+1]
       443B4208             cmp      r8d, dword ptr [rdx+8]
       732F                 jae      SHORT G_M5180_IG07
       488B3A               mov      rdi, bword ptr [rdx]
       4D63C0               movsxd   r8, r8d
       410FB6DA             movzx    rbx, r10b
       42881C07             mov      byte  ptr [rdi+r8], bl
       41C1FA08             sar      r10d, 8
       41FFC3               inc      r11d
       4183FB04             cmp      r11d, 4
       448BC6               mov      r8d, esi
       7CD8                 jl       SHORT G_M5180_IG04
```
There's also an extra mov in the span outer loop.

**Comment by AndyAyersMS** [comment]
There's always going to be at least one extra load somewhere in span case for these kinds of examples. Span args are passed implicitly by reference, so given a pointer to a span, you need two indirections to access the elements of a span, while given a pointer to an array, you only need one indirection to access the elements of an array.

The hope is that the jit will recognize that that the fields of a span are generally a local invariant (say by promoting the span as in #10453) so the extra load can be hoisted out of frequently executed code.

In the example above this happens for `src` but (for some reason) doesn't happen for `dst`. So the jit reloads both the length and pointer fields from `dst` on each iteration. Presumably the jit thinks the write via the pointer potentially aliases these fields.

FWIW I see an extra outer loop load for the array case (the array length of `dst`), not the span case. Seems like this could have been hoisted further to sit just below the outer loop zero trip check, but for some reason it wasn't.

cc @JosephTremoulet

**Comment by JosephTremoulet** [comment]
Right, due to the implicit-by-ref nature of the calling convention, we see `dst` as a byref to a span, and avoid hoisting those loads out of the loop for fear that `dst[b++]` could be overwriting parts of the span itself (that the byref parameter `dst` is a pointer to).

It's true that callers passing implicit-by-ref parameters are required to make an isolated copy for the callee to see and pass a pointer to that, but due to some details of our current program modeling, the compiler "knows" this early and "forgets" it by the time we're doing the optimizations that would eliminate these redundant loads.

The tool we have to do better for these cases is to (early on, when we "know" the rules) make a local copy of `dst` and rewrite the uses of `dst` as uses of the copy, then (later, when we've "forgotten" the special rules for implicit byrefs) we'd know that `dst[b++]` isn't overwriting that local copy.  We use a [hueristic](https://github.com/dotnet/coreclr/blob/ca3f2b3da810f6e8866a3a451a78ebed8dad03c8/src/jit/morph.cpp#L17981) to decide whether making the copy is worth it or not, and don't do this transformation when the static number of appearances of the parameter is too low.  Here, since there's just a single use of `dst` in the method, we avoid introducing the copy and thus don't keep track of the fact that `dst[b++]` can't overwrite it.

So this is another case where having static profile estimates early would help, since [then the heuristic could be sensitive to them](https://github.com/dotnet/coreclr/pull/10453#issuecomment-297837847) -- here the single use is in a nested loop, so we'd be more inclined to make the copy.

Also, of course, this is a case where it would behoove us to "remember" the special aliasing rules for implicit-by-ref parameters throughout compilation (or at least through more of it).

> FWIW I see an extra outer loop load for the array case (the array length of dst), not the span case. Seems like this could have been hoisted further to sit just below the outer loop zero trip check, but for some reason it wasn't.

When we run the invariant code motion analysis, we haven't yet optimized away the zero-trip-test for the inner loop, so we think that hoist would be speculative and thus don't do it.

**Comment by AndyAyersMS** [comment]
Interesting. So `src` has two appearances and that warrants promotion? Or do you really see 3 appearances for `src` (two length field references and a pointer reference) and 2 for `dst` (length and pointer?). 

Wonder if we should just be more aggressive with span and not worry as much about the downside?

**Comment by JosephTremoulet** [comment]
We see 3 `src` appearances and 2 `dst` appearances.

> Wonder if we should just be more aggressive with span and not worry as much about the downside?

It's a good question... would be nice to have a sizable corpus of code using spans to evaluate such a change against.

**Comment by stephentoub** [comment]
> It's a good question... would be nice to have a sizable corpus of code using spans to evaluate such a change against.

Working on it :smile:
https://github.com/dotnet/corefx/issues?q=is%3Aissue+is%3Aopen+span+label%3Aapi-approved

**Comment by AndyAyersMS** [comment]
Consider addressing for 2.1.

**Comment by AndyAyersMS** [comment]
Just to be clear, the thing we're considering here is to be more aggressive about copying a by-ref like argument like a `Span<T>` that is passed implicitly by-ref to a temp local, and then rewriting the in-method appearances to refer to the temp instead of the arg. We know this is safe because callers must have passed an unaliased reference.

This transformation typically allows the temp local to be promoted.

The heuristic is trying to weigh whether we should pay some some extra prolog cost (for the copy to the temp) to get some extra in-body savings (for the appearances). The cost of the copy is proportional to the number of fields. The cost of the appearances can vary and we don't have a great model for it. But we can count them easily enough.

The current heuristic is that the number of appearances of the arg has to be greater than the number of fields in the type of the arg.  So since `Span<T>` has 2 fields it needs 3 occurrences to trigger promotion.

**Comment by AndyAyersMS** [comment]
Went back and looked at the original and the ~20% perf hit is still there.

It's easy enough to mitigate this specific case in the importer, as the expansion of `Span.getItem` invokes `gtClone` on the index tree with the "complexOK" flag set. But `gtClone` doesn't handle `GT_SUB` and so as is we end up spilling to a temp.

Adding cloning support for `GT_SUB` fixes the above case but only fires in 4 places in jit-diffs -- and in two of them it causes small regressions. But one can probably justify the change on the grounds that if `GT_ADD` is considered cheap, so should `GT_SUB`.

**Comment by AndyAyersMS** [comment]
Circling back to the promotion -- I don't see a big benefit so far from being more aggressive with byref-likes, though it could be I'm not looking at the right examples.

Most times with Span, by being more aggressive we're just altering the way a copy gets made, as a copy has two uses. The use in a loop variant seems rare, and even then the memory references aren't so bad; in fact I have it clocking as a bit faster than loading the span in to registers before the loop.

If the byref span is used more heavily then this then it was already getting promoted.

So I'm thinking we should table the more aggressive promotion idea for now, pending either a stronger signal about the importance of the use (from say weighted ref data) or other examples that show why it's needed.

I'll send a follow-up PR extending `gtClone` since while the impact is small it makes the jit's behavior more uniform and predictable.

**Comment by AndyAyersMS** [comment]
One more note on this. While it's fairly rare for methods to modify their parameters, it does happen.  IIRC, writes to gc fields of struct parameters passed implicitly by-ref require checked write barriers. So we might consider screening for such cases and see if they're really as uncommon as I think. But it seems hard to imagine we'd find such a case without some subsequent use -- otherwise the side effects are pointless -- so likely they're already being promoted with the current heuristic.

**Comment by AndyAyersMS** [comment]
Original code example perf is fixed via #16128. The second example perf seems to be a wash despite the extra in-loop memory loads.

So I'm going to close this one out.
